### ISSUE TITLE
Decouple memory buffers from columns.

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#40222]
**Use case**

There are the following possibilities:
- allocate buffers in shared memory;
- allocate buffers in hugetlbfs;
- allocate buffers in GPU memory or other non-uniform memory;
- borrow a buffer for manipulation by external code;
- make a buffer refcounted and allow references to a sub-range of it;
- the buffer can own or not own its memory allocation;
- make a buffer readonly with mprotect and clear the flag on destruction (note: we already do it in debug build);
- calculate a checksum, verify and assert in destructor.
- make it compressed or offloaded and lazy reconstruct the contents on page fault;
- use a cache for column's data, but make its contents usable directly in-place, without memcpy;
- attach the data from Arrow directly into a column;
- attach the data from NumPy directly into a column;
- in some cases - avoid memcpy from ReadBuffer into Column while reading.

**Describe the solution you'd like**

A Buffer interface with (char *, size_t) as public members, but with virtual destructor and some virtual methods (create a new empty buffer of another size, realloc if possible, etc).

A factory can be used to create Buffers of various type, and already created Buffer (even empty) can be used as a factory to create more Buffers of the same type.

A parameterization of PODArray can be used to provide a version of this class using polymorphic Buffer, while a normal PODArray remains without any virtual methods.

PODArray, Column, and DataType (for the methods that create Columns) can take the Buffer object to create the Buffers of needed types.

The overhead should be only on construction and destruction, but for reading/writing the contents it is just a naked span of memory (zero overhead).

**Comment by zeromem** [comment]
Hi, any update on this topic? interested in 'cache for column', 'avoid memcpy', and want to know how to implement this. is there any in-progress MR ?

**Comment by alexey-milovidov** [comment]
@zeromem No, this task is not in progress. I have discussed it briefly with the team and also with @amosbird 
@amosbird said that he is interested, but I don't think that the work started.

**Comment by devwq** [comment]
Hi, any update on this topic? :)

**Comment by alexey-milovidov** [comment]
@devwq This task is still not in progress. We discussed it several times within the team in the context of other tasks. This task is what we should do, sooner or later.

**Comment by al13n321** [comment]
If this adds reference-counted copy-on-write buffers, can they replace and simplify the COW column stuff? There would be normal non-cow ColumnVector (whatever it'll be called) like in the good old days, but it'll point to a memory buffer that's copied-on-write.

(Not sure how unsharing would work if there's no concept of immutable vs mutable column pointers. But I'm also not sure why explicit column unsharing is needed in the first place. Can't we just say that e.g. IColumn::insertRangeFrom() implementation would cow-unshare its buffer, mutate it, then make it shared again? Then no one outside those methods would have to worry about COW: no IColumn::mutate()/shallowMutate()/MutableColumnCallback, Chunk::mutateColumns(), and all that.)

**Comment by alexey-milovidov** [comment]
Yes, this is a good idea.

**Comment by zhanglistar** [comment]
@alexey-milovidov Any update about this isssue? And lazy column?  Thanks a lot.

**Comment by alexey-milovidov** [comment]
@zhanglistar I discussed it with a few engineers, but we didn't plan it yet.

**Comment by alexey-milovidov** [comment]
First experiment is here: https://github.com/ClickHouse/ClickHouse/pull/77139
